server/location/index.js

```diff
@@ -0,0 +1,175 @@
+
+      // the only permissable objects are
+      // . latLng (leaflet.js)
+      // . or req.query object containing at least one of pc, pc7, pc8, pcd, latlong as properties
+      const fullPc = location.pc7 || location.pc8 || location.pc || location;
+      if (isFullPostcode(fullPc))
+        try {
+    console.log('got pc:',fullPc);
+          let specificity;
+          const info = await locationInfoFromPostcode(fullPc)
+    // console.log(Object.keys(info));
+          if (info.region)
+            specificity = 1
+          if (isPostcodeDistrict(info.pcd))
+            specificity = 5
+          if (info.parliamentary_constituency)
+            specificity = 6
+          if (toLatLong(info.latLong) || isFullPostcode(info.pc || info.pc7 || info.pc8))
+            specificity = 10
+          // TODO: use GSS bits here from GSS converter
+
+          Object.assign (result, info, {specificity} );
+
+          // job done, don't examine other cases
+          cache.put(fullPc, result, noVerify);
+          console.log('Successful lookup, got: ', info);
+          return result
+        }
+        catch (err) {
+          if (err.message.endsWith('404')) {
+            if (typeof location==='string')
+              location = districtFromFullPostcode(fullPc)
+            else
+              location.pcd = location.pcd || districtFromFullPostcode(fullPc)
+          }
+          console.log(err);
+        }
+
+      if (result.specificity<5 && isPostcodeDistrict(location.pcd || location)) {
+        // TODO: use cache, google, etc, to get some sense from partial postcode if it is real
+        Object.assign (result, {
+          pcd : location.pcd || location,
+          specificity : 5
+        });
+        //  cache but don't return - we may get a better result
+      }
+
+      // TODO: other string types, or, eg, query.city
+
+      // NB using lowercase latlong, as it's a querystring
+      location = location.latlong || location.pcd || (typeof location === 'string') ? location : null
+
+      if (latLongFrom4dpLatLongString(location)) {
+        const info = locationInfoFromLatLong(location)
+        /// process info
+        Object.assign ( result, info, { specificity:10 } )
+        cache.put( location, result, noVerify );
+        return result
+      }
+      if (options.useGoogle) {
+        // If useGoogle==true, use our API credits to try to get a more specific location from Google API
+        cache.put (location, result);
+      }
+
+      if (options.useTwitterContext) {
+        // Use context from eg user's tweets to attempt to guess location
+        // user data, eg screen name is received in options.useTwitterContext object
+      }
+      return result
+    }) ;
+
+  let bestLocation = (await Promise.all(possibles))
+    .filter (x=>x)
+    .sort ((a,b) => (b.specificity||0) - (a.specificity||0) )      // Specificity takes priority over input order
+
+// console.log('results:',bestLocation );
+
+  return bestLocation[0];
```

Comment:
what if there are multiple locations with the same specificity?

Reply:
They begin ordered by precedence they are passed in, and if the specificty of a pair or subset is equal, they retain this order.
Net result is they end up ordered by specificity and then by passed order, which is the aim.